storage/sink/kafka: Allow configuring auto created topic for Kafka sinks

rjobanp opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

  * This PR adds a known-desirable feature. Fixes https://github.com/MaterializeInc/materialize/issues/23794
(except for the schema registry compatibility policy - I will do that in a separate PR since it's a different statement block)


<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]


    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

maddyblue commented on src/sql-parser/src/parser.rs:
```diff
@@ -4399,9 +4421,44 @@ impl<'a> Parser<'a> {
+    fn parse_option_map<T, F>(
+        &mut self,
+        mut f: F,
+    ) -> Result<Option<BTreeMap<String, T>>, ParserError>
```
I'm a little concerned about the differences between this and `parse_map`:
- only strings as keys
- returns a btree instead of a vec of pairs

Only supporting string keys seems maybe ok because this needs to interact with `Value` which isn't an `Expr`. But then returning a `Vec<(String, T)>` would be needless if some other step is going to turn it into a btree anyway. Ok maybe this is fine then?

benesch replied:
> Ok maybe this is fine then?

Yeah, seems fine to me!

Also, there's no reason we'd ever need to call this with a value type of anything besides `WithOptionValue<T>`, so I'd actually prefer if we dropped the `mut f: F` argument and had this return a `Result<Option<BTreeMap<String, WithOptionValue<Raw>>>>`.

maddyblue reviewed (commented):
Adapter parts lgtm.

morsapaes reviewed (approved):
The syntax looks good to me, so +1 as a member of the SQL council. Thank you, @rjobanp!

def- commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3575,6 +3578,20 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
+                let len = values.len();
+                for (i, (key, value)) in values.iter().enumerate() {
+                    f.write_str("'");
+                    f.write_str(key.as_str());
```
No escaping?

benesch replied:
Yeah, def want `display::escape_single_quote_string` here!

def- reviewed (changes requested):
Thanks for adding tests. I'll add a platform-check and think of some failure cases.

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -2415,11 +2416,32 @@ impl<'a> Parser<'a> {
+                    KafkaSinkConfigOptionName::ProgressGroupIdPrefix
+                }
+                TOPIC => {
+                    self.expect_keywords(&[TOPIC, REPLICATION, FACTOR])?;
```
Looks like you have a double `TOPIC` here. Think you just want `[REPLICATION, FACTOR]`.

rjobanp replied:
Ah good catch, will update the sql parser tests so they would have caught this!

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -4399,9 +4421,44 @@ impl<'a> Parser<'a> {
+                }
+                let key = match self.next_token() {
+                    Some(Token::String(s)) => s,
+                    Some(Token::Ident(s)) => s.into_inner(),
```
Preference to just keep this to strings. Option values have to conflate idents and strings for PostgreSQL compatibility, but we can avoid that wart in map keys at least.

benesch commented on src/sql-parser/tests/testdata/ddl:
```diff
@@ -730,6 +730,14 @@ CREATE SINK foo IN CLUSTER c FROM bar INTO KAFKA CONNECTION baz (TOPIC = 'topic'
 CreateSink(CreateSinkStatement { name: Some(UnresolvedItemName([Ident("foo")])), in_cluster: Some(Unresolved(Ident("c"))), if_not_exists: false, from: Name(UnresolvedItemName([Ident("bar")])), connection: Kafka { connection: Name(UnresolvedItemName([Ident("baz")])), options: [KafkaSinkConfigOption { name: Topic, value: Some(Value(String("topic"))) }], key: None, headers: Some(Ident("header_column ...
 
+parse-statement
+CREATE SINK foo FROM bar INTO KAFKA CONNECTION baz (TOPIC 'topic', TOPIC CONFIG MAP['test_op' => 'val_1', 'test_op2' => 'val_2'], TOPIC REPLICATION FACTOR 4)
```
Seems worth testing the `PROGRESS TOPIC REPLICATION FACTOR` and `TOPIC PARTITION COUNT` options here too.

benesch commented on src/storage/src/sink/kafka.rs:
```diff
@@ -864,13 +855,7 @@ async fn determine_sink_progress(
-            // on Kafka connections.
-            replication_factor: -1,
-            cleanup_policy: TopicCleanupPolicy::Compaction,
-        },
```
❤️

benesch commented on src/storage/src/sink/kafka.rs:
```diff
@@ -615,15 +614,7 @@ fn sink_collection<G: Scope<Timestamp = Timestamp>>(
-                                ms: Some(-1),
-                                bytes: Some(-1),
-                            },
-                        },
```
❤️

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -3056,6 +3091,19 @@ fn kafka_sink_builder(
+            topic_config,
+        },
+        progress_topic_options: KafkaSinkTopicOptions {
+            // We only allow configuring the progress topic replication factor for now.
```
I'd also say "For correctness, the partition count MUST be one and for performance the compaction policy MUST be enabled."

benesch commented on src/storage-types/src/sinks.rs:
```diff
@@ -382,6 +383,36 @@ impl KafkaSinkCompressionType {
 
+#[derive(Arbitrary, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Default)]
+pub struct KafkaSinkTopicOptions {
+    /// The replication factor for the topic.
+    /// If `None`, the broker default will be used.
+    pub replication_factor: Option<i32>,
+    /// The number of partitions to create.
+    /// If `None`, the broker default will be used.
+    pub partition_count: Option<i32>,
```
What happens if these are `Some(-1)`? It seems like `Option<NonNeg<i32>>` would be safer, in that you don't need to worry about the -1 case representing the same thing as the `None` case.

benesch commented on src/storage-client/src/sink.rs:
```diff
@@ -213,55 +169,46 @@ pub async fn ensure_kafka_topic(
     // Newer versions of Kafka can instead send create topic requests with -1 and have this happen
     // behind the scenes, but this is unsupported and will result in errors on pre-2.4 Kafka.
-    if partition_count == -1 || replication_factor == -1 {
+    if partition_count.is_none() || replication_factor.is_none() {
```
Commented on below, but the use of `Option<i32>` admits two ways to represent "broker default" (-1 and `None`), which means code might disagree on which to use.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -3044,6 +3066,19 @@ fn kafka_sink_builder(
     };
 
+    let mut topic_config = topic_config.unwrap_or_default();
+    // If the user did not provide a cleanup policy for the topic, default to infinite
+    // retention.
+    if !topic_config.contains_key("cleanup.policy") {
+        topic_config.insert("cleanup.policy".to_string(), "delete".to_string());
+        if topic_config.contains_key("retention.bytes") || topic_config.contains_key("retention.ms")
+        {
+            sql_bail!("Cannot specify retention.bytes or retention.ms without cleanup.policy");
+        }
+        topic_config.insert("retention.bytes".to_string(), "-1".to_string());
+        topic_config.insert("retention.ms".to_string(), "-1".to_string());
+    }
```
I know the old code did this, but it is at odds with our documentation: https://materialize.com/docs/sql/create-sink/kafka/#automatic-topic-creation

> If the specified Kafka topic does not exist, Materialize will attempt to create it using the broker’s default number of partitions, default replication factor, default compaction policy, and default retention policy.

I'm inclined to make the code match the docs here and just _not_ set anything about `cleanup.policy` or `retention.{bytes,ms}` by default.

benesch commented on src/sql/src/plan/with_options.rs:
```diff
@@ -619,3 +622,26 @@ impl TryFromValue<WithOptionValue<Aug>> for ClusterScheduleOptionValue {
+    fn name() -> String {
+        format!("map of string to {}", V::name())
+    }
+}
```
This implementation went down easier than I would have expected!

benesch commented on test/testdrive/kafka-sink-topic-config.td:
```diff
@@ -0,0 +1,65 @@
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+  ENVELOPE DEBEZIUM
+
+# TODO: Figure out how to verify the replication factor was applied successfully
+# default cleanup policy should be indefinite retention
```
Supposedly the length of this array in the topic partition metadata tells you how many replicas that part has: https://docs.rs/rdkafka/latest/rdkafka/metadata/struct.MetadataPartition.html#method.replicas

benesch reviewed (commented):
Nice work! Left a few inline comments but nothing structural.

Can you work with @morsapaes to get the docs updated? It's a small enough change that I think we should do the docs update inline in this PR. The `{{< warn-if-unreleased >}}` flag will keep users from getting confused before the change hits prod.

benesch commented:
+1 from me on behalf of the SQL council as well.

def- reviewed (changes requested):
Found a panic:
```sql
CREATE SINK invalid_topic_options
IN CLUSTER quickstart
FROM v1
INTO KAFKA CONNECTION kafka_conn (
  TOPIC 'testdrive-kafka-sink-errors-4184745076',
  TOPIC CONFIG MAP['x''' => 'y''']
)
KEY(f1)
FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
ENVELOPE DEBEZIUM;
```
```
thread 'coordinator' panicked at src/adapter/src/catalog/builtin_table_updates.rs:1219:33:
create_sql cannot be invalid: CREATE SINK "materialize"."public"."invalid_topic_options" IN CLUSTER [u1] FROM [u1 AS "materialize"."public"."v1"] INTO KAFKA CONNECTION [u2 AS "materialize"."public"."kafka_conn"] (TOPIC = 'testdrive-kafka-sink-errors-4184745076', TOPIC CONFIG = MAP['x'' => 'y''']) KEY ("f1") FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION [u3 AS "materialize"."public"."csr_conn"] ENVELOPE DEBEZIUM
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:72:14
   2: <mz_adapter::catalog::state::CatalogState>::pack_sink_update::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/builtin_table_updates.rs:1219:33
   3: <core::result::Result<alloc::vec::Vec<mz_sql_parser::parser::StatementParseResult>, mz_sql_parser::parser::ParserStatementError>>::unwrap_or_else::<<mz_adapter::catalog::state::CatalogState>::pack_sink_update::{closure#0}>
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/result.rs:1431:23
   4: <mz_adapter::catalog::state::CatalogState>::pack_sink_update
             at ./home/deen/git/materialize3/src/adapter/src/catalog/builtin_table_updates.rs:1219:14
   5: <mz_adapter::catalog::state::CatalogState>::pack_item_update
             at ./home/deen/git/materialize3/src/adapter/src/catalog/builtin_table_updates.rs:529:21
   6: <mz_adapter::catalog::Catalog>::transact_op::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:1070:46
   7: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact_op::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
   8: <mz_adapter::catalog::Catalog>::transact_op::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:499:5
   9: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:473:14
  10: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  11: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:434:5
  12: <mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:395:10
  13: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  14: <mz_adapter::catalog::Catalog>::transact::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/catalog/transact.rs:331:5
  15: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:529:14
  16: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  17: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:181:5
  18: <mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:119:71
  19: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  20: <mz_adapter::coord::Coordinator>::catalog_transact_conn::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:113:5
  21: <mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:76:14
  22: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  23: <mz_adapter::coord::Coordinator>::catalog_transact::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/ddl.rs:69:5
  24: <mz_adapter::coord::Coordinator>::sequence_create_sink::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/sequencer/inner.rs:987:70
  25: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::sequence_create_sink::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  26: <mz_adapter::coord::Coordinator>::sequence_create_sink::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/sequencer/inner.rs:925:5
  27: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/sequencer.rs:203:72
  28: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  29: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/future/future.rs:123:9
  30: <mz_adapter::coord::Coordinator>::message_purified_statement_ready::{closure#0}::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/message_handler.rs:549:85
  31: <mz_adapter::coord::Coordinator>::message_purified_statement_ready::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/message_handler.rs:463:5
  32: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord/message_handler.rs:93:66
  33: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_message::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  34: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/future/future.rs:123:9
  35: <mz_adapter::coord::Coordinator>::serve::{closure#0}
             at ./home/deen/git/materialize3/src/adapter/src/coord.rs:2837:48
  36: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/future/future.rs:123:9
  37: <tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  38: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/park.rs:282:63
  39: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:107:5
  40: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:73:5
  41: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/park.rs:282:31
  42: <tokio::runtime::context::blocking::BlockingRegionGuard>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/blocking.rs:66:9
  43: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/handle.rs:310:22
  44: tokio::runtime::context::runtime::enter_runtime::<<tokio::runtime::handle::Handle>::block_on<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}, ()>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/runtime.rs:65:16
  45: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/handle.rs:309:9
  46: mz_adapter::coord::serve::{closure#0}::{closure#2}
             at ./home/deen/git/materialize3/src/adapter/src/coord.rs:3422:21
```

def- reviewed (changes requested):
If you create an invalid configuration map, how do we want to have this handled? Currently the `CREATE SINK` still succeeds, but then we have errors in our logs:
```
testdrive-materialized-1          | cluster-u1-replica-u1: 2024-06-12T04:57:00.840085Z  INFO mz_storage::healthcheck: Broadcasting suspend-and-restart command because of Some((Kafka, Stalled { error: "Error creating topic testdrive-kafka-config-unknown-475791216 for sink: Admin operation error: InvalidConfig (Broker: Configuration is invalid)", hint: None, should_halt: true })) after 30s delay
testdrive-kafka-1                 | [2024-06-12 04:57:00,737] INFO [Admin Manager on Broker 1]: Error processing create topic request CreatableTopic(name='testdrive-kafka-config-unknown-475791216', numPartitions=1, replicationFactor=1, assignments=[], configs=[CreateableTopicConfig(name='abc', value='def'), CreateableTopicConfig(name='cleanup.policy', value='delete'), CreateableTopicConfig(name='retention.bytes', value='-1'), CreateableTopicConfig(name='retention.ms', value='-1')]) (kafka.server.ZkAdminManager)
testdrive-kafka-1                 | org.apache.kafka.common.errors.InvalidConfigurationException: Unknown topic config name: abc
```
I'll add a test for that so that we at least have that documented.

benesch commented:
> If you create an invalid configuration map, how do we want to have this handled? Currently the `CREATE SINK` still succeeds, but then we have errors in our logs:

Yeah, that's expected. We can't know whether the topic config is valid until we actually go to create the topic. You _should_ see the error in `mz_kafka_sink_statuses` though?

def- commented:
Yes, that's how I wrote the test for now, then we can keep it as is (except for the panic).

rjobanp commented:
Alright I think I've addressed all the comments on this PR - including a light stab at the documentation updates. @morsapaes Feel free to amend that as needed!

def- reviewed (approved):
Thank you for the fixes, I triggered another nightly: https://buildkite.com/materialize/nightly/builds/8076 (Edit: Green)

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -60,12 +60,16 @@ _item&lowbar;name_ | The name of the source, table or materialized view you want
+`COMPRESSION TYPE`                  | `text`              | Default: `none`. The type of compression to apply to messages before they are sent to Kafka: `none`, `gzip`, `snappy`, `lz4`, or `zstd`.
+`TRANSACTIONAL ID PREFIX`           | `text`              | The prefix of the transactional ID to use when producing to the Kafka topic.<br>Default: `materialize-{REGION ID}-{CONNECTION ID}-{SINK ID}`
+`PROGRESS GROUP ID PREFIX`          | `text`              | The prefix of the consumer group ID to use when reading from the progress topic.<br>Default: `materialize-{REGION ID}-{CONNECTION ID}-{SINK ID}`
+`TOPIC REPLICATION FACTOR`          | `int`               | The replication factor to use when creating the Kafka topic (if applicable).<br>Default: Broker's default
+`TOPIC PARTITION COUNT`             | `int`               | The partition count to use when creating the Kafka topic (if applicable).<br>Default: Broker's default
+`PROGRESS TOPIC REPLICATION FACTOR` | `int`               | The partition count to use when creating the Kafka [progress topic](#exactly-once-processing) (if applicable).<br>Default: Broker's default
+`TOPIC CONFIG`                      | `map[text => text]` | Any topic-level configs to use when creating the Kafka topic (if applicable).<br>See [the kafka docs](https://kafka.apache.org/documentation/#topicconfigs) for available options.
```
```suggestion
`TOPIC REPLICATION FACTOR`          | `int`               | The replication factor to use when creating the Kafka topic (if the Kafka topic does not already exist).<br>Default: Broker's default
`TOPIC PARTITION COUNT`             | `int`               | The partition count to use when creating the Kafka topic (if the Kafka topic does not already exist).<br>Default: Broker's default
`PROGRESS TOPIC REPLICATION FACTOR` | `int`               | The partition count to use when creating the Kafka [progress topic](#exactly-once-processing) (if the Kafka topic does not already exist).<br>Default: Broker's default
`TOPIC CONFIG`                      | `map[text => text]` | Default: empty<br>Any topic-level configs to use when creating the Kafka topic (if the Kafka topic does not already exist).<br>See the [Kafka documentation](https://kafka.apache.org/documentation/#topicconfigs) for available configs.
```

Took me a second to realize what you meant by "if applicable". I think it's worth being a bit more verbose but fully explicit. Also I think worth saying the default for the map is empty.

benesch replied:
Can you add `{{< warn-if-unreleased-inline "v0.104" }}` to each of these?

benesch commented on doc/user/content/sql/create-sink/kafka.md:
```diff
@@ -358,16 +362,21 @@ Consider using the Debezium envelope if:
 Materialize will attempt to create it with a single partition, the broker's
 default replication factor, compaction enabled, and both size- and time-based
-retention disabled.
+retention disabled. The replication factor can be overriden using the
```
```suggestion
retention disabled. The replication factor can be overridden using the
```

benesch reviewed (approved):
Very nice! Thanks for knocking this one out.
